### PR TITLE
🌱 Add build arg to override builder image

### PR DESCRIPTION
**Pull request by ykakarap**
**What this PR does / why we need it**:
This PR allows users to override the builder image.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

**Review by fabriziopandini — COMMENTED** [review]
/lgtm

**Comment by fabriziopandini** [comment]
@vincepri 
If the argument is not set it will use the default (=golang:1.16.5)

**Comment by vincepri** [comment]
If we move the argument to the Makefile as a variable, we can reuse it for all the other instances where we declare the Go image version:

- https://github.com/kubernetes-sigs/cluster-api/blob/master/Makefile#L375
- https://github.com/kubernetes-sigs/cluster-api/blob/master/Makefile#L544

File: Makefile [hunk 1]
```diff
@@ -20,6 +20,9 @@ SHELL:=/usr/bin/env bash
 
 .DEFAULT_GOAL:=help
 
+GOLANG_VERSION ?= 1.16.5
```
Comment (vincepri) [inline Makefile:23]:
```suggestion
GO_VERSION ?= 1.16.5
```

File: Makefile [hunk 2]
```diff
@@ -20,6 +20,9 @@ SHELL:=/usr/bin/env bash
...
+BUILDER_IMAGE ?= docker.io/library/golang:$(GOLANG_VERSION)
```
Comment (vincepri) [inline Makefile:24]:
```suggestion
GO_CONTAINER_IMAGE ?= docker.io/library/golang:$(GOLANG_VERSION)
```

File: test/infrastructure/docker/Makefile [hunk 3]
```diff
@@ -19,6 +19,9 @@ ROOT = ../../..
 
 .DEFAULT_GOAL:=help
 
+GOLANG_VERSION ?= 1.16.5
+BUILDER_IMAGE := golang:$(GOLANG_VERSION)
```
Comment (vincepri) [inline test/infrastructure/docker/Makefile:23]: Same as above

**Review by vincepri — APPROVED** [review]
/approve
/lgtm

**Pull request merged**